dosbox-staging/dosbox-staging
# The great DOS_File/DOS_Drive Refactor (plus some timestamp improvements)

# Description

Several changes here.  Probably recommend reviewing by commit because I've got broken down decently well.

General overview:

- DOS_Files array moved from raw pointer to `unique_ptr`
- DOS_Drive moved from "owning" vectors of `unique_ptr` with lots of raw pointers everywhere to `shared_ptr`.
- Improvements to timestamp handling to better match MS-DOS behavior


## Related issues

Fixes #620


# Manual testing

Wrote test programs in x86 assembly and ran on MS-DOS 6.22 to confirm behavior of various combinations of DOS interrupts to see when timestamps get updated.  Opened multiple handles to the same file. Used the get date interrupt to verify when the timestamp gets updated after write, truncate, create, set date interrupt, etc.

Tested a few games with save files to verify they still work:  Crystal Caves, Doom, Secret of Monkey Island 2

Tested the case on #620 to verify that we now set the timestamp equal to Dosbox's BIOS rather than host time.

# Checklist

_Please tick the items as you have addressed them. Don't remove items; leave the ones that are not applicable unchecked._

I have:
